feat: wechatpay-apiv3/wechatpay-guzzle-middleware#17 敏感信息加解密功能

TheNorthMemory opened the pull request:
新增「加解密」功能， 测试校准@ PHP 7.3.11 with LibreSSL 2.8.3

- 上行加密@ `/v3/smartguide/guides`
- 下行解密@ `/v3/merchant-service/complaints`

备注：`OPENSSL_PKCS1_OAEP_PADDING` 自PHP7引入，此lib未在`PHP5`系列测试，使用时请注意PHP版本。

TheNorthMemory committed:
feat: wechatpay-apiv3/wechatpay-guzzle-middleware#17 敏感信息加解密功能

TheNorthMemory committed:
renew the sample usage of `Decrypt`

TheNorthMemory committed:
renew the mandatory header[Wechatpay-Serial] on README.md

TheNorthMemory committed:
typo `pem`, should be `PemUtil` on README.md

xy-peng commented on src/Util/SensitiveInfoCodec.php:
```diff
@@ -0,0 +1,139 @@
+ * </code>
+ *
+ * @package  WechatPay
+ * @author   James Zhang(https://github.com/TheNorthMemory)
```
处于很多考虑，我们不接受代码中留下具体作者的名字。希望能移除，谢谢。

TheNorthMemory replied:
行吧，我一并把 MediaUtil 也移除。

TheNorthMemory committed:
removed `@author`

xy-peng commented:
codec的说法可能不妥，通常用在base64/hex之类的场景。敏感信息加密准确的说应该是密码学，建议用crypto

tpirc3 commented on src/Util/SensitiveInfoCodec.php:
```diff
@@ -0,0 +1,138 @@
+     *
+     * @return SensitiveInfoCodec
+     */
+    private function encrypt($str) {
```
`encrypt`和`decrypt`这两个函数是否可以设置成public，并返回加密/解密后的字符串，这样除了结合setStage用于json序列化/反序列化，也可以快捷获取到加密/解密后的字符串用于调试和其他用途，用法更灵活些

TheNorthMemory replied:
实际在考虑把`encrypt`/`decrypt`这俩私有化是因为 `\JsonSerializable`，`json_encode` 会驱动 `jsonSerialize`，并且有 `__toString` 都可以捕获加密/解密后的字符串。

tpirc3 commented on src/Util/SensitiveInfoCodec.php:
```diff
@@ -0,0 +1,138 @@
+     *
+     * @return SensitiveInfoCodec
+     */
+    public function setStage($scenario) {
```
`setStage`的参数可否改成常量，或者校验一下`$scenario`是否为`encrypt`或`decrypt`之一，如果不校验的就使用的话，结合`__invoke`里面的动态函数执行可能会有意想不到的效果

TheNorthMemory replied:
这个可以有

tpirc3 commented on README.md:
```diff
@@ -137,6 +137,50 @@ try {
+        'json' => [
+            'business_code' => 'APL_98761234',
+            'contact_info'  => [
+                'contact_name'      => $encryptor('窃格瓦拉'),
```
这个example名字就太过不正式了😂，最好还是换一个吧

TheNorthMemory committed:
renamed to `SensitiveInfoCrypto`

- added the argument checking onto `setStage` function
- use the name of `Alice` onto README.md

TheNorthMemory commented:
有必要我再捋一下此类的设计目标吧：

- 之所以把 `encrypt` 和 `decrypt` 方法私有化，是和用法有关；若public化，实例在驱动加/解密时，现有处理方式，在实例内`message`会被干扰到，不改实现的话，出参就只能是字符串了，这就不好玩了；进而如果`trace` 的话，反倒是麻烦了（至少是要甄别源`json`敏感信息是不是本来就是加了密的还是被此 `crypto` 弄过）；

- 实例的方法，当静态方法用，其实和把实例直接当方法用(`__invoke`)没多大区别；

- 此类的 `__invoke` 中的 `(){}()` 语法糖，充分利用了语言特性，把纵向声明拉平成横向pipeline，抽象了  `encrypt` 和 `decrypt` 中的共有处理逻辑，让那俩方法中的处理变得单一清晰，进而可以让这俩方法私有掉，归一化入口入参；

- 封装的关键点其实在 `implements \JsonSerializable`，这个基类只有1个抽象方法，用途也很单一，即用在 `json_encode` 场景，这就为此 **中间件** 应用的业务单元，全程保持住了 _面向对象_（且这个对象内留有 **证书** 句柄及 **OAEP_PADDING** 声明），这些显性要素信息其实远远优于字符串，有助于**过程调试**；

TheNorthMemory committed:
[optim] on class header comment

TheNorthMemory committed:
toggle `$codec`->`$crypto` onto README.md

tpirc3 commented on src/Util/SensitiveInfoCrypto.php:
```diff
@@ -0,0 +1,152 @@
+     * @var int Equal to OPENSSL_PKCS1_OAEP_PADDING constant,
+     *          to prevent the fault error while the PHP_VERSION < 7.0.
+     */
+    const OPENSSL_PKCS1_OAEP_PADDING = 4;
```
针对`OPENSSL_PKCS1_OAEP_PADDING `我确认了下，至少PHP 5中都已经引入了该常量 https://github.com/php/php-src/blob/PHP-5.1/ext/openssl/openssl.c ，可以直接使用，我们只需要支持到PHP 5.5和PHP 7以上即可

TheNorthMemory replied:
好哒，容我找个环境试试，可以跑就直接常量走起了。

TheNorthMemory replied:
测试了下，在`PHP5.5`上用商户 `公共/私钥证书对` 跑了下第3个用例，可正常加/解密 `Alice`， 注释已调整优化。

TheNorthMemory committed:
OPENSSL_PKCS1_OAEP_PADDING constant was available since PHP 5.1

tpirc3 commented on src/Util/SensitiveInfoCrypto.php:
```diff
@@ -0,0 +1,152 @@
+     * @param resource|null $publicCert The offical public certificate resource
+     * @param resource|null $privateCert The merchant private certificate resource
+     */
+    public function __construct($publicCert, $privateCert = null) {
```
这里是否也可以对`$publicCert`和`$privateCert`的合法性进行校验，如果不合法的话抛一个`InvalidArgumentException`，因为使用者大概率会直接把 `$encryptor('value of field')`直接填到array里传给json_encode，不会进行判空，如果`$publicCert`加载的有问题，这个敏感字段的值就会变成空字符串传到API接口里

TheNorthMemory replied:
嗯，这个判断难搞，`(false, true)`, `(true, false)`, `(true, true)` 都有可能，只在 `(false,false)` 做异常提醒吧。

xy-peng replied:
个人觉得，整体分两层，下层提供encrypt/decrypt的裸方法，上层只针对json_encode提供封装，会更合适一些。毕竟大部分场景是需要加密上传，用目前的封装方式是非常好的。但是decode和decrypt并没有很好的结合，放在一起还容易用错。

TheNorthMemory replied:
完全同意不完美，这也是 `constructor`第二个参数可选的原因；加解密封放在一起，也是给开发者提供一份参考，加密用 `openssl_public_encrypt `，解密用 `openssl_private_decrypt`。

TheNorthMemory committed:
checking `is_resource` on Constructor

TheNorthMemory committed:
[optim] on usage comments
